src/Bicep.Core/Emit/ExpressionEmitter.cs

```diff
@@ -314,6 +314,46 @@ public void EmitObjectProperties(ObjectSyntax objectSyntax, ISet<string>? proper
             }
         }
 
+        public void EmitModuleParameterValue(SyntaxBase syntax)
+        {
+            if (syntax is InstanceFunctionCallSyntax instanceFunctionCall && string.Equals(instanceFunctionCall.Name.IdentifierName, "getSecret", StringComparison.Ordinal))
+            {
+                var objectSymbol = context.SemanticModel.GetSymbolInfo(instanceFunctionCall.BaseExpression);
+                if (objectSymbol is ResourceSymbol resource)
+                {
+                    writer.WritePropertyName("reference");
+                    writer.WriteStartObject();
+                    writer.WritePropertyName("keyVault");
+                    writer.WriteStartObject();
+
+                    writer.WritePropertyName("id");
+                    var keyVaultId = converter.GetFullyQualifiedResourceId(resource);
+                    var keyVaultIdSerialised = ExpressionSerializer.SerializeExpression(keyVaultId);
+                    writer.WriteValue(keyVaultIdSerialised);
+
+                    writer.WriteEndObject(); // keyVault
+
+                    writer.WritePropertyName("secretName");
+                    var secretName = converter.ConvertExpression(instanceFunctionCall.Arguments[0].Expression);
+                    var secretNameSerialised = ExpressionSerializer.SerializeExpression(secretName);
+                    writer.WriteValue(secretNameSerialised);
+
+                    if (instanceFunctionCall.Arguments.Length > 1)
+                    {
+                        writer.WritePropertyName("secretVersion");
+                        var secretVersion = converter.ConvertExpression(instanceFunctionCall.Arguments[1].Expression);
+                        var secretVersionSerialised = ExpressionSerializer.SerializeExpression(secretVersion);
+                        writer.WriteValue(secretVersionSerialised);
+                    }
+                    writer.WriteEndObject(); // reference
+
+                    return;
+                }
+                throw new InvalidOperationException("Cannot emit parameter's KeyVault secret reference.");
```

Comment:
Presumably this'll never actually happen in practice (e.g. you've already performed validation and blocked emitting). Mind confirming?

Reply:
Yes, it could be thrown only when function was called on something else than a Resource.